### PR TITLE
Replace clearState call-count tests with observable trace assertions (#529)

### LINKED ISSUE #529: 🟡 HOW-assertion counts clearState calls in common/episode_runner_test.ts
**Severity: medium** — `common/episode_runner_test.ts`, lines 90–119.

Two tests replace `creature.clearState` with a counting wrapper and assert on the number of invocations rather than on any observable result:

```ts
let clears = 0;
const original = creature.clearState.bind(creature);
creature.clearState = () => { clears++; original(); };
const result = runEpisode(creature, adapter, { maxSteps: 10 });
assertEquals(result.steps, 4);
assertEquals(clears, 4, "clearState called once per executed tick");
// sibling test: with { clearStatePerTick: false } asserts clears === 0
```

## Why this matters

This is a HOW-assertion (anti-pattern #1): the test stubs an internal helper of the unit under test and asserts it was called a specific number of times. It verifies the *mechanism* (`clearState` is invoked once per tick) instead of the *outcome* (the option produces correct per-tick state). If `runEpisode` were refactored to clear state via a different call, to batch the clears, or to reset inline without calling `clearState`, the observable episode result would be unchanged but these tests would fail. A bubble-sort→quicksort-style rewrite of the episode loop should not require touching these assertions; the coupling to internals is what obstructs refactoring. (The neighbouring `result.steps`/`result.trace` assertions in the same tests are good WHAT-assertions and should be kept.)

## Suggested fix

**Option (a) — assert observable behaviour.** Use a stub creature whose state visibly accumulates across ticks, then assert the *effect* of the flag rather than the call count: with `clearStatePerTick: true` the recorded per-tick trace shows no carry-over; with `false` it accumulates.

```ts
const onTrace  = runEpisode(creature, env, { clearStatePerTick: true }).trace;
const offTrace = runEpisode(creature, env, { clearStatePerTick: false }).trace;
assertNotEquals(onTrace, offTrace); // plus the concrete expected traces
```

**Option (b) — delete the call-count assertions.** Keep the `result.steps`/`result.trace` checks and drop the `clears` counting. If clearing state between ticks has no observable effect, there is nothing for a caller to depend on and the assertion adds no value.

### PR DESCRIPTION
**Pull request by VibeCoderST**
## Summary

Replaced two HOW-assertion tests in `common/episode_runner_test.ts` that stubbed
`creature.clearState` and asserted on the _number of invocations_ (`clears === 4`, `clears === 0`).
Counting an internal helper's calls verifies the _mechanism_, not the _outcome_ — a refactor of the
episode loop (batching clears, resetting inline, or clearing via a different call) would keep the
observable episode result identical yet break those assertions. This violates the project's
[Testing Philosophy](../../../AGENTS.md) (only "what" tests allowed). Closes #529.

The fix follows the issue's **Option (a)**: assert the _observable_ effect of the
`clearStatePerTick` flag through the returned `trace`, using a stub creature whose recurrent state
is visible in its activation output.

### Approach

The flag's only caller-visible effect is whether recurrent creature state carries across ticks. The
new tests model that directly with a tiny stub creature holding a single recurrent accumulator:

- `clearState()` resets the accumulator to `0`.
- `activate()` increments it and returns the running total as the output.

An `accumulatingWalker` adapter uses that output as its per-tick step delta, so any carry-over shows
up in the trace:

| `clearStatePerTick` | accumulator output | trace          |
| ------------------- | ------------------ | -------------- |
| `true` / default    | `1, 1, 1, 1`       | `[0,1,2,3,4]`  |
| `false`             | `1, 2, 3, 4`       | `[0,1,3,6,10]` |

```mermaid
flowchart LR
    subgraph cleared["clearStatePerTick: true (default)"]
        A0[0] -->|+1| A1[1] -->|+1| A2[2] -->|+1| A3[3] -->|+1| A4[4]
    end
    subgraph retained["clearStatePerTick: false"]
        B0[0] -->|+1| B1[1] -->|+2| B3[3] -->|+3| B6[6] -->|+4| B10[10]
    end
```

A refactor of the loop internals that preserves behaviour leaves these traces unchanged, so the
tests no longer obstruct refactoring while still covering both flag states.

## Evidence

Backend/test-only change — no UI to screenshot. Verified via `deno test`:

```
running 10 tests from ./common/episode_runner_test.ts
...
ok | 10 passed | 0 failed (10ms)
```

`deno fmt --check` and `deno lint` both pass on the modified file.

## Test Plan

Existing call-count tests replaced (documented modification — the issue's whole purpose);
neighbouring good `result.steps`/`result.trace` WHAT-assertions retained. New/changed tests in
`common/episode_runner_test.ts`:

- `runEpisode — clears recurrent state per tick by default (observable via trace)` — asserts the
  default and explicit `clearStatePerTick: true` both yield `[0,1,2,3,4]`.
- `runEpisode — clearStatePerTick=false retains recurrent state across ticks` — asserts the
  accumulating trace `[0,1,3,6,10]` and `finalState === 10`.
- `runEpisode — clearStatePerTick toggles the observable episode trace` — asserts the two traces
  differ, proving the flag has a real caller-visible effect.

---

🤖 Processed by: VibeCoderST
🏷️ Run id: `vibe-mpt9rgi5-24659c`

### CODE REVIEW

**Review by nleck — APPROVED** [review]

**Comment by VibeCoderST** [comment]
## CI failure needs human attention

**Why:** I reviewed the CI check failure (**Run quality checks**) but could not produce a code fix. The classifier indicates this requires a code change.

**Classifier reason:** code-fix pattern matched: ts2322
**Signals:**
- `text:ts2322`

**Next step:** Review the failing check and either push a fix or reply with guidance, then remove the `needs-human` label so the worker can retry.

---
🤖 Processed by: the worker

**Comment by VibeCoderST** [comment]
## CI fix: `Run quality checks` (deno check type error)

**Root cause.** The failure was in the `Run deno check (type checking)` step, not in this PR's clearState test changes. The automated dependency bump on this branch (commit `12bc232`, auto-bump #362) raised `@stsoftware/neat-ai` from **5.0.45 → 5.3.3**, but **5.3.3 removed the `CATEGORICAL_ERROR` cost function**. In 5.3.3 the `CostName` union is `CROSS_ENTROPY | MSE | MAE | MAPE | MSLE | HINGE`, so `deno check` failed:

```
TS2322 [ERROR]: Type '"CATEGORICAL_ERROR"' is not assignable to type
'"CROSS_ENTROPY" | "MSE" | "MAE" | "MAPE" | "MSLE" | "HINGE" | undefined'.
  mnist_classification/exploration_campaign.ts:541
  mnist_classification/mnist_classification.ts:683
```

I confirmed via a runtime probe against 5.3.3 that `CATEGORICAL_ERROR` is gone entirely (`Costs.getAvailableCosts()` → `[CROSS_ENTROPY, MSE, MAE, MAPE, MSLE, HINGE]`; `Costs.find("CATEGORICAL_ERROR")` throws `Unknown cost function`). The MNIST example fundamentally evolves with `CATEGORICAL_ERROR` (`MNIST_EVOLVE_COST_NAME`).

**Fix applied.** Reverted the unrelated neat-ai bump back to **5.0.45** in `deno.json` and `deno.lock` (the `Develop` base branch is also on 5.0.45). This restores compilation and keeps the PR focused on its actual scope (clearState observable-trace tests). Migrating MNIST off `CATEGORICAL_ERROR` to a supported cost would be a deliberate behavioural change across the example code, recorded artefacts, rust-scorer probing, docs and tests — out of scope here.

**Verification (local).**
- `deno check **/*.ts` — passes, zero errors.
- `deno check` on both previously-failing files — passes.
- `MNIST_EVOLVE_COST_NAME is registered in NEAT-AI` test — passes.

**Follow-up.** Opened stSoftwareAU/NEAT-AI-Examples#541 to track the underlying incompatibility (the auto-bump will keep re-breaking until neat-ai restores `CATEGORICAL_ERROR`, MNIST migrates to a supported cost, or the bump is pinned below the breaking version).

**Comment by nleck** [comment]
Please replace with CROSS_ENTROPY

**Comment by VibeCoderST** [comment]
## CI failure needs human attention

**Why:** I reviewed the CI check failure (**Run quality checks**) but could not produce a code fix. The classifier indicates this requires a code change.

**Classifier reason:** code-fix pattern matched: type error
**Signals:**
- `text:type error`
- `text:ts2322`

**Next step:** Review the failing check and either push a fix or reply with guidance, then remove the `needs-human` label so the worker can retry.

---
🤖 Processed by: the worker

**Comment by VibeCoderST** [comment]
## Automated Processing Failed (First Attempt)

The automated worker encountered an issue while processing this comment and will retry.

### Failure Details
Claude produced no output for 600 seconds (silence watchdog fired)

### What happens next?
- This comment has been marked for retry (confused reaction added)
- The worker will attempt to process it again on the next scan
- If it fails again, manual intervention will be required

**Comment by VibeCoderST** [comment]
## CI failure needs human attention

**Why:** I reviewed the CI check failure (**Run quality checks**) but could not produce a code fix. The classifier indicates this requires a code change.

**Classifier reason:** code-fix pattern matched: ts2322
**Signals:**
- `text:ts2322`

**Next step:** Review the failing check and either push a fix or reply with guidance, then remove the `needs-human` label so the worker can retry.

---
🤖 Processed by: the worker

**Comment by VibeCoderST** [comment]
## CI failure needs human attention

**Why:** I reviewed the CI check failure (**Run quality checks**) but could not produce a code fix. The classifier indicates this requires a code change.

**Classifier reason:** code-fix pattern matched: type error
**Signals:**
- `text:type error`
- `text:ts2322`

**Next step:** Review the failing check and either push a fix or reply with guidance, then remove the `needs-human` label so the worker can retry.

---
🤖 Processed by: the worker

**Comment by nleck** [comment]
Fix the below by replacing with CROSS_ENTROPY

Check xor_classification/xor_classification_test.ts
TS2322 [ERROR]: Type '"CATEGORICAL_ERROR"' is not assignable to type '"CROSS_ENTROPY" | "MSE" | "MAE" | "MAPE" | "MSLE" | "HINGE" | undefined'.
    options: { costName: MNIST_EVOLVE_COST_NAME },
               ~~~~~~~~
    at file:///home/runner/work/NEAT-AI-Examples/NEAT-AI-Examples/mnist_classification/exploration_campaign.ts:541:16

TS2322 [ERROR]: Type '"CATEGORICAL_ERROR"' is not assignable to type '"CROSS_ENTROPY" | "MSE" | "MAE" | "MAPE" | "MSLE" | "HINGE" | undefined'.
      costName: MNIST_EVOLVE_COST_NAME,
      ~~~~~~~~
    at file:///home/runner/work/NEAT-AI-Examples/NEAT-AI-Examples/mnist_classification/mnist_classification.ts:683:7

    The expected type comes from property 'costName' which is declared here on type 'NeatOptions'
      /** The name of the cost function to use. */ costName: CostName;
                                                   ~~~~~~~~
        at https://jsr.io/@stsoftware/neat-ai/5.3.3/src/config/NeatArguments.ts:50:3

Found 2 errors.

**Comment by VibeCoderST** [comment]
Claiming PR feedback comment 4586112231 for worker `GRQ-23-50575`.

**Comment by VibeCoderST** [comment]
Fixed — replaced the removed `CATEGORICAL_ERROR` cost with `CROSS_ENTROPY`, which resolves both TS2322 errors (`exploration_campaign.ts:541` and `mnist_classification.ts:683`). In neat-ai 5.3.3 `CostName` no longer includes `CATEGORICAL_ERROR`.

Changes:
- `mnist_classification/mnist_classification.ts`: `MNIST_EVOLVE_COST_NAME` is now `"CROSS_ENTROPY"`; updated the module and `targetError` doc comments to describe cross-entropy loss instead of the misclassification rate.
- `mnist_classification/mnist_classification_test.ts`: updated the registration test to assert `"CROSS_ENTROPY"` and to check that cross-entropy penalises a confident wrong prediction more than a confident correct one (the old `=== 1` assertion was specific to the misclassification-rate cost).
- `mnist_classification/evolve_integration_test.ts`: relaxed `assertValidEvolveResult` — cross-entropy loss is non-negative but unbounded above, so the old `bestError <= 1` upper bound (valid only for the `[0,1]` misclassification rate) no longer holds.

`deno check` now passes for all `mnist_classification/*` and `xor_classification/*` files, and the MNIST unit + evolveDir integration tests pass (43 + 7 green).

Out of scope (pre-existing, not caused by this fix):
- `xor_classification/xor_classification_test.ts` "solves XOR" and the intermittent `crispr_injection_test.ts` case fail independently of this change — confirmed by stashing my edits and re-running. They are seedless, stochastic evolution tests affected by neat-ai 5.3.3 behavioural changes; XOR uses MSE, not the cost I changed. These should be tracked separately rather than bundled into this type-error fix.
- The native-scorer probes in `quality.sh` / `ensure_neat_ai_native_scorer*` / CI still reference `CATEGORICAL_ERROR`; those guard the optional rust scorer (currently unavailable, JS fallback in use) and are a separate concern.

**Comment by VibeCoderST** [comment]
Failed to fix CI failure (**Run quality checks**): Claude timed out after 3600 seconds

**Comment by VibeCoderST** [comment]
## Fixed the `Run quality checks` failure

The failing job (`Run unit tests with coverage`) reported two test failures:

- `cart_pole/cart_pole_test.ts:272` — *champion generalises to unseen perturbed initial states*: `got 345.3`, needed `≥384`.
- `xor_classification/xor_classification_test.ts:124` — *solves XOR and grows hidden neurons*: `error=0.25` after 402 generations (stuck on the MSE plateau).

### Root cause

Neither failing file is touched by this PR. The trigger is the auto-bump of `@stsoftware/neat-ai` from `5.0.45` → `5.3.3` (an internal package, so the `Deno Dependency Auto-Bump` workflow bumps it past the quarantine gate on every PR push — reverting the pin is not durable). I confirmed this locally: the XOR test passes on `5.0.45` and fails deterministically on `5.3.3`. Version 5.3.3 changed the evolutionary/structural-mutation dynamics, so the previous population budgets no longer converge robustly:

- **XOR**: a 30-creature population no longer grows the hidden neuron XOR needs within 400 generations (the default cost is still `MSE`, and the champion stalls at the 0.25 plateau). A 50-creature population (the documented `DEFAULT_EVOLVE_OPTIONS` value) solves in ~76 generations with `error≈0.005` — a 10× margin under the 0.05 threshold.
- **cart-pole**: a 60-creature population produced champions that solved the training trials (`bestScore≥480`, which is why the sibling test passed) but generalised marginally to the held-out perturbed starts. Cart-pole physics is chaotic, so a marginal champion is sensitive to cross-platform float divergence — locally it scored 486, on CI 345. A 100-creature population finds a comfortably robust champion that balances the full episode (generalisation 500/500 locally) with ample margin above the 384 floor.

### Changes

- `xor_classification/xor_classification_test.ts` — happy-path test population `30 → 50`.
- `cart_pole/cart_pole.ts` — `DEFAULT_EVOLVE_OPTIONS.populationSize` `60 → 100`.

No tests were disabled, skipped, or weakened — both still assert the genuine behaviour (XOR solved + hidden neuron grown; cart-pole champion generalises above `0.8 × SOLVED_THRESHOLD`).

### Verification (local)

- `cart_pole/cart_pole_test.ts` — **24 passed / 0 failed**.
- `xor_classification/xor_classification_test.ts` — **23 passed / 0 failed**.
- `deno fmt --check`, `deno lint`, `deno check` — clean on both changed files.

`deno.json` / `deno.lock` are intentionally untouched (left at `5.3.3`).
